src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1996,55 +2003,107 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+                                        receiverSyntax, receiverSyntax, receiver, methodName, rightArity: 0,
+                                        typeArgumentsSyntax: default(SeparatedSyntaxList<TypeSyntax>), typeArguments: default(ImmutableArray<TypeSymbol>),
+                                        invoked: true, diagnostics: diagnostics);
 
-            if (deconstructMethod.ParameterCount != numCheckedVariables)
-            {
-                Error(diagnostics, ErrorCode.ERR_DeconstructWrongParams, boundRHS.Syntax, deconstructMethod, numCheckedVariables);
-                return null;
-            }
+                memberAccess = CheckValue(memberAccess, BindValueKind.RValueOrMethodGroup, diagnostics);
+                memberAccess.WasCompilerGenerated = true;
+
+                if (memberAccess.Kind != BoundKind.MethodGroup)
+                {
+                    return MissingDeconstruct(receiver, assignmentSyntax, numCheckedVariables, diagnostics, out outPlaceholders, receiver);
+                }
 
-            if (deconstructMethod.Parameters.Any(p => p.RefKind != RefKind.Out))
+                // After the overload resolution completes, the last step is to coerce the arguments with inferred types.
+                // That step returns placeholder (of correct type) instead of the outVar nodes that were passed in as arguments.
+                // So the generated invocation expression will contain placeholders instead of those outVar nodes.
+                // Those placeholders are also recorded in the outVar for easy access below, by the `SetInferredType` call on the outVar nodes.
+                bag = DiagnosticBag.GetInstance();
+                BoundExpression result = BindMethodGroupInvocation(
+                                            receiverSyntax, receiverSyntax, methodName, (BoundMethodGroup)memberAccess, analyzedArguments, bag, queryClause: null,
+                                            allowUnexpandedForm: true);
+
+                result.WasCompilerGenerated = true;
+
+                if (bag.HasAnyErrors())
+                {
+                    return MissingDeconstruct(receiver, assignmentSyntax, numCheckedVariables, diagnostics, out outPlaceholders, result);
+                }
+                else
+                {
+                    diagnostics.AddRange(bag);
+                }
+
+                // Verify all the parameters (except "this" for extension methods) are out parameters
+                var deconstructMethod = ((BoundCall)result).Method;
+                var parameters = deconstructMethod.Parameters;
+                for (int i = (deconstructMethod.IsExtensionMethod ? 1 : 0); i < parameters.Length; i++)
+                {
+                    if (parameters[i].RefKind != RefKind.Out)
+                    {
+                        return MissingDeconstruct(receiver, assignmentSyntax, numCheckedVariables, diagnostics, out outPlaceholders, result);
```

Comment:
Yes, in the tests that involve optional or params parameters. For example, `DeconstructMethodHasOptionalParam`.